### ISSUE TITLE
Improving static type checking results

### ISSUE THREAD

**Issue by Paebbels** [edaa-org/pyEDAA.UCIS#11]
@por3bski I wanted to ask you if you maybe have an idea how to improved the static type checking results for the code when it comes to using `lxml`.

I added `lxml-stubs` to the dependencies on `dev` branch, but still the results are not so good with respect to types from `lxml`.  
See [installation of lxml-stubs](https://github.com/edaa-org/pyEDAA.UCIS/runs/4947176046?check_suite_focus=true#step:4:23)

Here is the mypy output: https://github.com/edaa-org/pyEDAA.UCIS/runs/4947176046?check_suite_focus=true#step:5:10  
Here is the matching HTML report: https://edaa-org.github.io/pyEDAA.UCIS/typing/index.html

I'm also using `lxml` in other places, but never investigated what to improve in such a situation.

**Comment by por3bski** [comment]
The main problem are types unions returned from lxml functions.
For example xpath(...) can return str, int, list, bool etc.

I have to use "assert" with isinstance() to narrow union types:

```
for scopeNode in scopes:
	assert isinstance(scopeNode, etree._Element) <----------

	if scopeNode.get("type").startswith("DU_"):
		continue
```
				
If this is accetable I can prepare PR in next few days.
I do not guarantee that it fixes every lxml error because lxml-stubs package is not complete:

> These type annotations were initially included in typeshed, but lxml's annotations are still incomplete and have therefore been extracted from typeshed to avoid unintentional false positive results.

**Comment by Paebbels** [comment]
I think I would prefer:
```py
for scopeNode in scopes:
  if isinstance(scopeNode, etree._Element):  # <----------
    # some code here

  elif isinstance(scopeNode, etree.xxxxx):   # <----------
    # some code here

  else:
    raise ....                               # <----------
```

as a coding pattern instead of assert statements.

----
Because you have direct write access to this repository you can branch again from `dev`. You automatically get CI fully enabled. This generates all the reports like I linked them above. Then you can investigate differences and see how `lxml` behaves with some changes. If your commits improve the reports issues, you might go on and fix the next problems.

Note: You can also link your PR to this issues, so this issue gets closed when the PR is merged to the `main` branch :).

**Comment by umarcor** [comment]
Implemented in #12 and merged to *main* in #10.
